test/test_utils/ec2.py

```diff
@@ -257,6 +269,43 @@ def launch_instance(
     return response["Instances"][0]
 
 
+@retry(
+    reraise=True,
+    stop=stop_after_delay(30 * 60),  # Keep retrying for 30 minutes
+    wait=wait_random_exponential(min=60, max=5 * 60),  # Retry after waiting 1-5 minutes
+)
+def launch_instances_with_retry(
+    ec2_resource, availability_zone_options, ec2_create_instances_definition
+):
+    """
+    Helper function to launch EC2 instances with retry capability, to allow multiple attempts
+    when facing instance capacity issues.
+    :param ec2_resource: boto3 EC2 Service Resource object
+    :param availability_zone_options: list of availability zones in which to try to run instances
+    :param ec2_create_instances_definition: dict of parameters to pass to
+        ec2_resource.create_instances
+    :return: list of EC2 Instance Resource objects for instances launched
+    """
+    instances = None
+    if availability_zone_options:
+        error = None
+        for a_zone in availability_zone_options:
+            ec2_create_instances_definition["Placement"] = {"AvailabilityZone": a_zone}
+            try:
+                instances = ec2_resource.create_instances(**ec2_create_instances_definition)
+                if instances:
+                    break
```

Comment:
nit: We could have returned the instances from here directly.

Reply:
We could return from here, but I'm trying to ensure that we return from the end of the function, in case we need to add some additional logic before the return.

For these kinds of functions that could go wrong in multiple ways, I also thought it would be better to have just a single point to return from, to make it easier to debug.